tests/test_histogram.py

```diff
@@ -110,3 +117,85 @@ def test_compare_histograms_optimized_vs_general():
     assert_allclose(hist_ghc_root['sum_hessians'], hist_ghc_root['count'])
     assert_allclose(hist_gc['sum_hessians'], np.zeros(n_bins))  # unused
     assert_allclose(hist_ghc['sum_hessians'], hist_ghc['count'])
+
+
+def test_unrolled_equivalent_to_naive():
+    # Make sure the different unrolled histogram computations give the same
+    # results as the naive one.
+    rng = np.random.RandomState(42)
+    n_samples = 10
+    n_bins = 5
+    sample_indices = np.arange(n_samples).astype(np.uint32)
+    binned_feature = rng.randint(0, n_bins - 1, size=n_samples, dtype=np.uint8)
+    ordered_gradients = rng.randn(n_samples).astype(np.float32)
+    ordered_hessians = np.ones(n_samples, dtype=np.float32)
+
+    hist_gc_root = _build_gc_root_histogram_unrolled(n_bins, binned_feature,
+                                                     ordered_gradients)
+    hist_ghc_root = _build_ghc_root_histogram_unrolled(n_bins, binned_feature,
+                                                       ordered_gradients,
+                                                       ordered_hessians)
+    hist_gc = _build_gc_histogram_unrolled(n_bins, sample_indices,
+                                           binned_feature,
+                                           ordered_gradients)
+    hist_ghc = _build_ghc_histogram_unrolled(n_bins, sample_indices,
+                                             binned_feature,
+                                             ordered_gradients,
+                                             ordered_hessians)
+
+    hist_naive = _build_ghc_histogram_naive(n_bins, sample_indices,
+                                            binned_feature, ordered_gradients,
+                                            ordered_hessians)
+
+    for hist in (hist_gc_root, hist_ghc_root, hist_gc, hist_gc, hist_ghc):
+        assert_array_equal(hist['count'], hist_naive['count'])
+        assert_allclose(hist['sum_gradients'], hist_naive['sum_gradients'])
+    for hist in (hist_ghc_root, hist_ghc):
+        assert_allclose(hist['sum_hessians'], hist_naive['sum_hessians'])
+    for hist in (hist_gc_root, hist_gc):
+        assert_array_equal(hist['sum_hessians'], np.zeros(n_bins))
+
+
+def test_hist_subtraction():
```

Comment:
Please parametrize this test to test both the constant hessian and a non-constant hessian case:

```python
@pytest.mark.parametrize("constant_hessian", [True, False]):
def test_hist_subtraction(constant_hessian):
    ...
    if constant_hessian:
        ordered_hessians = np.ones(n_samples, dtype=np.float32)
    else:
        ordered_hessians = rng.lognormal(size=n_samples).astype(np.float32)
    ...
```

and when `constant_hessian` is `True` use the `no_hessian` variants of the `_build_histogram` methods both for the parent and the sibling data scanning computations.